docs/tutorials/wiki2/src/tests/tutorial/security.py

```diff
@@ -8,33 +9,50 @@
 from . import models
 
 
-class MyAuthenticationPolicy(AuthTktAuthenticationPolicy):
+class MySecurityPolicy:
+    def __init__(self, secret):
+        self.authtkt = AuthTktCookieHelper(secret)
+        self.acl = ACLHelper()
+
+    def authenticated_identity(self, request):
+        identity = self.authtkt.identify(request)
+        if identity is None:
+            return None
+
+        userid = identity['userid']
+        user = request.dbsession.query(models.User).get(userid)
+        return user
+
     def authenticated_userid(self, request):
-        user = request.user
+        user = self.authenticated_identity(request)
         if user is not None:
             return user.id
 
+    def remember(self, request, userid, **kw):
+        return self.authtkt.remember(request, userid, **kw)
+
+    def forget(self, request, **kw):
+        return self.authtkt.forget(request, **kw)
+
+    def permits(self, request, context, permission):
+        principals = self.effective_principals(request)
+        return self.acl.permits(context, principals, permission)
+
     def effective_principals(self, request):
         principals = [Everyone]
-        user = request.user
+        user = self.authenticated_identity(request)
         if user is not None:
             principals.append(Authenticated)
-            principals.append(str(user.id))
+            principals.append('u:' + str(user.id))
             principals.append('role:' + user.role)
         return principals
 
-def get_user(request):
-    user_id = request.unauthenticated_userid
-    if user_id is not None:
-        user = request.dbsession.query(models.User).get(user_id)
-        return user
-
 def includeme(config):
     settings = config.get_settings()
-    authn_policy = MyAuthenticationPolicy(
-        settings['auth.secret'],
-        hashalg='sha512',
-    )
-    config.set_authentication_policy(authn_policy)
-    config.set_authorization_policy(ACLAuthorizationPolicy())
-    config.add_request_method(get_user, 'user', reify=True)
+
+    config.set_csrf_storage_policy(CookieCSRFStoragePolicy())
+    config.set_default_csrf_options(require_csrf=True)
+
+    config.set_security_policy(MySecurityPolicy(settings['auth.secret']))
+    config.add_request_method(
+        lambda request: request.authenticated_identity, 'user', property=True)
```

Comment:
Why not `reify` to add caching?

Reply:
`request.user` is not the source of truth. Caching is hard - if you cache something that is not the source, you're doing it wrong. Even if I rewrote the policy to use `request.user` everywhere, `request.authenticated_identity` is still an api, and would not be cached. Caching needs to be done in `authenticated_identity` by the policy, if we're going to do it. In this case it is already basically cached because there are no database queries on subsequent calls thanks to SQLAlchemy's identity map.